Bluetooth: host: Fix ATT error code on GATT authentication procedure

sjanc opened the pull request:
“Insufficient Encryption” shall be returned only if link is not encrypted and proper LTK is present, otherwise “Insufficient Authentication” shall be used.

Core Specification 5.4 Vol. 3 Part C 10.3.1

"If neither an LTK nor an STK is available, the service request shall be rejected with the error code
“Insufficient Authentication”.
Note: When the link is not encrypted, the error code “Insufficient Authentication” does not indicate that MITM protection is required.
If an LTK or an STK is available and encryption is required (LE security mode 1) but encryption is not enabled, the service request shall be rejected with the error code “Insufficient Encryption”."

This was affecting GAP/SEC/AUT/BV-11-C qualification test case.

sjanc committed:
Bluetooth: host: Fix ATT error code on GATT authentication procedure

“Insufficient Encryption” shall be returned only if link is not
encrypted and proper LTK is present, otherwise “Insufficient
Authentication” shall be used.

Core Specification 5.4 Vol. 3 Part C 10.3.1

"If neither an LTK nor an STK is available, the service
request shall be rejected with the error code
“Insufficient Authentication”.
Note: When the link is not encrypted, the error code
“Insufficient Authentication” does not indicate that
MITM protection is required.
If an LTK or an STK is available and encryption is
required (LE security mode 1) but encryption is not
enabled, the service request shall be rejected with
the error code “Insufficient Encryption”."

This was affecting GAP/SEC/AUT/BV-11-C qualification test case.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

Thalley commented:
Hmm, I thought we handled this in `l2cap_check_security` which seems very similar to the change introduced here, just at a lower (L2CAP) level. 

Might something be missing from `l2cap_check_security`, rather than implementing this in GATT?

sjanc commented:
l2cap_check_security() is for LE COC, not related to GATT

sjanc commented:
yeap, BTW I'm not yet sure why some EDTT tests are failing though  (trying to setup this locally now)

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3090,6 +3090,23 @@ uint16_t bt_gatt_get_mtu(struct bt_conn *conn)
 	return bt_att_get_mtu(conn);
 }
 
+#if defined(CONFIG_BT_SMP)
+static bool ltk_present(struct bt_conn *conn)
+{
+	struct bt_keys *keys = conn->le.keys;
+
+	if (keys) {
+		if (conn->role == BT_HCI_ROLE_CENTRAL) {
+			return keys->keys & (BT_KEYS_LTK_P256 | BT_KEYS_PERIPH_LTK);
+		} else {
+			return keys->keys & (BT_KEYS_LTK_P256 | BT_KEYS_LTK);
+		}
+	}
+
+	return false;
+}
+#endif /* CONFIG_BT_SMP */
```
Suggest to make this an internal function in `conn.c` so that it can be used here in GATT and modify `l2cap_check_security` to use this function too.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3109,26 +3126,47 @@ uint8_t bt_gatt_check_perm(struct bt_conn *conn, const struct bt_gatt_attr *attr
-		if (!conn->encrypt) {
-			return BT_ATT_ERR_INSUFFICIENT_ENCRYPTION;
+		if (mask & BT_GATT_PERM_LESC_MASK) {
+			struct bt_keys *keys = conn->le.keys;
```
```suggestion
			const struct bt_keys *keys = conn->le.keys;
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3090,6 +3090,23 @@ uint16_t bt_gatt_get_mtu(struct bt_conn *conn)
+#if defined(CONFIG_BT_SMP)
+static bool ltk_present(struct bt_conn *conn)
+{
+	struct bt_keys *keys = conn->le.keys;
```
```suggestion
	const struct bt_keys *keys = conn->le.keys;
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3090,6 +3090,23 @@ uint16_t bt_gatt_get_mtu(struct bt_conn *conn)
 }
 
+#if defined(CONFIG_BT_SMP)
+static bool ltk_present(struct bt_conn *conn)
```
```suggestion
static bool ltk_present(const struct bt_conn *conn)
```

Thalley reviewed (commented):
The implementation looks good, just a few suggestions

sjanc committed:
bluetooth: host: Add common helper for checking LTK presence

Both L2CAP and GATT have same requirements with regards to error code
on no encryption when LTK is or isn't present.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc committed:
tests/bsim: Disable some broken EDTT GATT tests

Disable tests that are failing due to invalid test implementation.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

sjanc commented:
Added commit that (for now) disables failing EDTT tests, no code changes
